tests/Mqtt5ClientTest.cpp

```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
+
+    const int MESSAGE_NUMBER = 10;
+    std::atomic<int> client_messages(0);
+    bool client1_received = false;
+    bool client2_received = false;
+
+    std::vector<int> receivedMessages;
+    for (int i = 0; i < MESSAGE_NUMBER; i++)
+    {
+        receivedMessages.push_back(0);
+    }
+
+    Aws::Iot::Mqtt5ClientBuilder *subscribe_builder =
+        Aws::Iot::Mqtt5ClientBuilder::NewMqtt5ClientBuilderWithMtlsFromPath(
+            mqtt5TestVars.m_hostname_string,
+            mqtt5TestVars.m_certificate_path_string.c_str(),
+            mqtt5TestVars.m_private_key_path_string.c_str(),
+            allocator);
+    ASSERT_TRUE(subscribe_builder);
+
+    std::promise<void> client_received;
+    auto onMessage_client1 = [&](const PublishReceivedEventData &eventData) -> int {
+        String topic = eventData.publishPacket->getTopic();
+        if (topic == TEST_TOPIC)
+        {
+            ByteCursor payload = eventData.publishPacket->getPayload();
+            String message_string = String((const char *)payload.ptr, payload.len);
+            int message_int = atoi(message_string.c_str());
+            ASSERT_TRUE(message_int < MESSAGE_NUMBER);
+            ++receivedMessages[message_int];
+            client1_received = true;
+
+            bool exchanged = false;
+            int desired = 11;
+            int tested = 10;
+            client_messages++;
+            exchanged = client_messages.compare_exchange_strong(tested, desired);
+            if (exchanged == true)
+            {
+                client_received.set_value();
+            }
+        }
+        return 0;
+    };
+    subscribe_builder->WithPublishReceivedCallback(onMessage_client1);
+
+    Aws::Iot::Mqtt5ClientBuilder *subscribe_builder2 =
+        Aws::Iot::Mqtt5ClientBuilder::NewMqtt5ClientBuilderWithMtlsFromPath(
+            mqtt5TestVars.m_hostname_string,
+            mqtt5TestVars.m_certificate_path_string.c_str(),
+            mqtt5TestVars.m_private_key_path_string.c_str(),
+            allocator);
+    ASSERT_TRUE(subscribe_builder2);
+
+    auto onMessage_client2 = [&](const PublishReceivedEventData &eventData) -> int {
+        String topic = eventData.publishPacket->getTopic();
+        if (topic == TEST_TOPIC)
+        {
+            ByteCursor payload = eventData.publishPacket->getPayload();
+            String message_string = String((const char *)payload.ptr, payload.len);
+            int message_int = atoi(message_string.c_str());
+            ASSERT_TRUE(message_int < MESSAGE_NUMBER);
+            ++receivedMessages[message_int];
+            client2_received = true;
+
+            bool exchanged = false;
+            int desired = 11;
+            int tested = 10;
+            client_messages++;
+            exchanged = client_messages.compare_exchange_strong(tested, desired);
+            if (exchanged == true)
+            {
+                client_received.set_value();
+            }
+        }
+        return 0;
```

Comment:
Trivial nit: the callbacks for both subscribers are basically the same, so it's worth introducing a standalone function/lambda.